test/unittest/unittest_nntrainer_models.cpp

```diff
@@ -212,63 +210,61 @@ void NodeWatcher::read(std::ifstream &in) {
 
 void NodeWatcher::verifyWeight(const std::string &error_msg) {
   for (unsigned int i = 0; i < expected_weights.size(); ++i) {
-    verify(node->weightAt(i).getVariable(), expected_weights[i].getVariable(),
-           error_msg + " " + node->weightAt(i).getName() + " weight");
+    verify(node.layer->weightAt(i).getVariable(),
+           expected_weights[i].getVariable(),
+           error_msg + " " + node.layer->weightAt(i).getName() + " weight");
   }
 }
 
 void NodeWatcher::verifyGrad(const std::string &error_msg) {
   for (unsigned int i = 0; i < expected_weights.size(); ++i) {
-    auto weight = node->weightAt(i);
+    auto weight = node.layer->weightAt(i);
     if (weight.getTrainable()) {
       verify(weight.getGradient(), expected_weights[i].getGradient(),
              error_msg + " " + weight.getName() + " grad");
     }
   }
 }
 
-nntrainer::sharedConstTensors
-NodeWatcher::forward(nntrainer::sharedConstTensors in, int iteration) {
+void NodeWatcher::forward(int iteration) {
   std::stringstream ss;
-  ss << "forward failed at " << node->getName() << " at iteration "
+  ss << "forward failed at " << node.layer->getName() << " at iteration "
      << iteration;
   std::string err_msg = ss.str();
 
-  nntrainer::sharedConstTensors out = node->forwarding(in);
-  verify(*out[0], expected_output, err_msg + " at output");
-  return out;
+  std::vector<nntrainer::Tensor> out = node.layer->getHidden();
+
+  verify(out[0], expected_output, err_msg + " at output");
 }
 
 nntrainer::sharedConstTensors
 NodeWatcher::lossForward(nntrainer::sharedConstTensors pred,
                          nntrainer::sharedConstTensors answer, int iteration) {
   std::stringstream ss;
-  ss << "loss failed at " << node->getName() << " at iteration " << iteration;
+  ss << "loss failed at " << node.layer->getName() << " at iteration "
+     << iteration;
   std::string err_msg = ss.str();
 
   nntrainer::sharedConstTensors out =
-    std::static_pointer_cast<nntrainer::LossLayer>(node)->forwarding(pred,
-                                                                     answer);
+    std::static_pointer_cast<nntrainer::LossLayer>(node.layer)
+      ->forwarding(pred, answer);
 
   return out;
 }
 
-nntrainer::sharedConstTensors
-NodeWatcher::backward(nntrainer::sharedConstTensors deriv, int iteration,
-                      bool should_verify) {
+void NodeWatcher::backward(int iteration, bool should_verify) {
   std::stringstream ss;
-  ss << "backward failed at " << node->getName() << " at iteration "
+  ss << "backward failed at " << node.layer->getName() << " at iteration "
      << iteration;
   std::string err_msg = ss.str();
 
-  nntrainer::sharedConstTensors out = node->backwarding(deriv, iteration);
+  std::vector<nntrainer::Tensor> out = node.layer->getGradient();
+
   if (should_verify) {
+    verify(out[0], expected_dx, err_msg);
```

Comment:
Any reason for the order change?
IMO gradient should be verified beforehand,

Reply:
fixed.